src/librustc_borrowck/borrowck/mir/dataflow/sanity_check.rs

```diff
@@ -0,0 +1,193 @@
+    // performed by `build_sets`.)
+
+    let blocks = mir.all_basic_blocks();
+    'next_block: for bb in blocks {
+        let bb_data = mir.basic_block_data(bb);
+        let &repr::BasicBlockData { ref statements,
+                                    ref terminator,
+                                    is_cleanup: _ } = bb_data;
+
+        let (args, span) = if let Some(repr::Terminator { ref kind, span, .. }) = *terminator {
+            if let repr::TerminatorKind::Call { func: ref oper, ref args, .. } = *kind
+            {
+                if let repr::Operand::Constant(ref func) = *oper
+                {
+                    if let ty::TyFnDef(def_id, _, &ty::BareFnTy { abi, .. }) = func.ty.sty
+                    {
+                        let name = tcx.item_name(def_id);
+                        if abi == Abi::RustIntrinsic || abi == Abi::PlatformIntrinsic {
+                            if name.as_str() == "rustc_peek" {
+                                (args, span)
+                            } else {
+                                continue;
+                            }
+                        } else {
+                            continue;
+                        }
+                    } else {
+                        continue;
+                    }
+                } else {
+                    continue;
+                }
+            } else {
+                continue;
+            }
+        } else {
+            continue;
+        };
+        assert!(args.len() == 1);
+        let peek_arg_lval = match args[0] {
+            repr::Operand::Consume(ref lval @ repr::Lvalue::Temp(_)) => {
+                lval
+            }
+            repr::Operand::Consume(_) => {
+                bug!("dataflow::sanity_check cannot feed a non-temp to rustc_peek.");
+            }
+            repr::Operand::Constant(_) => {
+                bug!("dataflow::sanity_check cannot feed a constant to rustc_peek.");
+            }
+        };
+
+        let mut entry = results.0.sets.on_entry_set_for(bb.index()).to_owned();
+        let mut gen = results.0.sets.gen_set_for(bb.index()).to_owned();
+        let mut kill = results.0.sets.kill_set_for(bb.index()).to_owned();
+
+        let move_data = flow_ctxt.move_data();
+
+        // Emulate effect of all statements in the block up to (but
+        // not including) the assignment to `peek_arg_lval`. Do *not*
+        // include terminator (since we are peeking the state of the
+        // argument at time immediate preceding Call to `rustc_peek`).
+
+        let mut sets = super::BlockSets { on_entry: &mut entry[..],
+                                          gen_set: &mut gen[..],
+                                          kill_set: &mut kill[..] };
+        // Unfortunately if we just re-do the same thing that dataflow does, then
+        // it will always appear like Lvalues are initialized; e.g. in
+        // a case like:
+        //
+        // <bitset maps var1 to 0>
```

Comment:
Oh you're right; I didn't update that comment correctly (it was originally authored for a different version of `rustc_peek` that looked like `rustc_peek(var1)` (instead of `rustc_peek(&var1)`), and I erroneously translated its content without thinking about what the actual translation of `rustc_peek(&var1)` is).

(The original version of `rustc_peek` had a more fragile API, where one would write `rustc_peek(expr)` (instead of `rustc_peek(&expr)`).)

----

The main point I'm trying to make is this: even with the corrected translation that you wrote:

```
tmp1 = &var0;
tmp0 = rustc_peek(tmp1) -> bb1;
```

The normal (naive) dataflow GEN/KILL sets are going to say that `tmp1` is initialized. If the semantics of `rustc_peek` were a naive "look directly at your argument's dataflow bit state", then `rustc_peek` would query the state of `tmp1`, which would always indicate that it has been initialized under the naive rules.

So I had written a comment about a work-around I was using for the naive system.

----

With the new API the story is somewhat related, but the example I gave in the comment isn't helping explain things.

I will update it.